qiskit_optimization/problems/substitute_variables.py

```diff
@@ -0,0 +1,308 @@
+    ) -> QuadraticProgram:
+        """Substitutes variables with constants or other variables.
+
+        Args:
+            src: a quadratic program to be substituted.
+
+            constants: replace variable by constant
+                e.g., {'x': 2} means 'x' is substituted with 2
+
+            variables: replace variables by weighted other variable
+                need to copy everything using name reference to make sure that indices are matched
+                correctly. The lower and upper bounds are updated accordingly.
+                e.g., {'x': ('y', 2)} means 'x' is substituted with 'y' * 2
+
+        Returns:
+            An optimization problem by substituting variables with constants or other variables.
+            If the substitution is valid, `QuadraticProgram.status` is still
+            `QuadraticProgram.Status.VALID`.
+            Otherwise, it gets `QuadraticProgram.Status.INFEASIBLE`.
+
+        Raises:
+            QiskitOptimizationError: if the substitution is invalid as follows.
+                - Same variable is substituted multiple times.
+                - Coefficient of variable substitution is zero.
+        """
+        self._src = src
+        self._dst = QuadraticProgram(src.name)
+        self._subs_dict(constants, variables)
+        results = [
+            self._variables(),
+            self._objective(),
+            self._linear_constraints(),
+            self._quadratic_constraints(),
+        ]
+        if any(not r for r in results):
+            self._dst._status = QuadraticProgram.Status.INFEASIBLE
+        return self._dst
+
+    @staticmethod
+    def _feasible(sense: ConstraintSense, rhs: float) -> bool:
+        """Checks feasibility of the following condition
+        0 `sense` rhs
+        """
+        # I use the following pylint option because `rhs` should come to right
+        # pylint: disable=misplaced-comparison-constant
+        if sense == ConstraintSense.EQ:
+            if 0 == rhs:
+                return True
+        elif sense == ConstraintSense.LE:
+            if 0 <= rhs:
+                return True
+        elif sense == ConstraintSense.GE:
+            if 0 >= rhs:
+                return True
+        return False
+
+    @staticmethod
+    def _replace_dict_keys_with_names(op, dic):
+        key = []
+        val = []
+        for k in sorted(dic.keys()):
+            key.append(op.variables.get_names(k))
+            val.append(dic[k])
+        return key, val
+
+    def _subs_dict(self, constants, variables):
+        # guarantee that there is no overlap between variables to be replaced and combine input
+        subs = {}  # type: Dict[Union[int, str], Tuple[str, float]]
+        if constants is not None:
+            for i, v in constants.items():
+                # substitute i <- v
+                i_2 = self._src.get_variable(i).name
+                if i_2 in subs:
+                    raise QiskitOptimizationError(
```

Comment:
Again something we can consider changing, if you have time/inclination, over to using f-Strings, which we are trying to standardize on using rather than the older .format()

Reply:
Sure. I will replace them with f-strings